File: elixir/square-root/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand), do: radix(radicand, 1, radicand)

  defp radix(aim, bot, top) do
    middle = (bot + top) / 2
    square = middle * middle

    cond do
      square == aim -> middle
      square > aim -> radix(aim, bot, middle)
      square < aim -> radix(aim, middle, top)
    end
  end
end

File: java/square-root/src/main/java/SquareRoot.java
public class SquareRoot {
    public int squareRoot(int radicand) {
        return this.findSquareRoot(radicand, 1, radicand);
    }

    private int findSquareRoot(int radicand, int bot, int top) {
        final int middle = (bot + top) / 2;
        final int square = middle * middle;

        if (square == radicand) {
            return middle;
        }
        if (square > radicand) {
            return findSquareRoot(radicand, bot, middle);
        }
        return findSquareRoot(radicand, middle, top);
    }

}
